staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go

```diff
@@ -121,6 +132,197 @@ const (
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Specification of the desired behavior of the AdmissionPolicy.
+	// +optional
+	Spec AdmissionPolicySpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ValidatingAdmissionPolicyList is a list of ValidatingAdmissionPolicy.
+type ValidatingAdmissionPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// List of ValidatingAdmissionPolicy.
+	Items []ValidatingAdmissionPolicy `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// AdmissionPolicySpec is the specification of the desired behavior of the AdmissionPolicy.
+type AdmissionPolicySpec struct {
+	//ParamSource specifies the kind of resources used to parameterize this policy
+	// +optional
+	ParamSource ParamSource `json:"ParamSource,omitempty" protobuf:"bytes,1,rep,name=ParamSource"`
+
+	// MatchConstraints specifies what resources this policy is designed to validate.
+	// The AdmissionPolicy cares about an validation if it matches _any_ Constriant.
+	// However, in order to prevent ValidatingAdmissionPolicy
+	// from putting the cluster in a state which cannot be recovered from without completely
+	// disabling the process, ValidatingAdmissionPolicy is never called
+	// on admission requests for AdmissionPolicySpec objects.
+	// +optional
+	MatchConstraints []RuleWithOperations `json:"MatchConstraints,omitempty" protobuf:"bytes,2,rep,name=matchConstraints"`
+
+	// Validations contain CEL expressions which is used to apply the validation.
+	// Required.
+	Validations []Validation `json:"Validations,omitempty" protobuf:"bytes,3,rep,name=Validations"`
+}
+
+// ParamSource is a tuple of Group Kind and Version.
+type ParamSource struct {
+	// APIGroup is the API group the resources belong to. '*' is all groups.
+	// Required.
+	APIGroup string `json:"apiGroup,omitempty" protobuf:"bytes,1,rep,name=apiGroup"`
+
+	// APIVersion is the API version the resources belong to. '*' is all versions.
+	// Required.
+	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,2,rep,name=apiVersion"`
+
+	// APIKind is the API kind the resources belong to. '*' is all versions.
+	// Required.
+	APIKind string `json:"apiKind,omitempty" protobuf:"bytes,3,rep,name=apiKind"`
+}
+
+// Validation specified the CEL expression which is used to apply the validation.
+type Validation struct {
+	// The name of the validation rule.
+	// Required.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// The Expression for validation.
+	// Required.
+	Expression string `json:"expression" protobuf:"bytes,2,opt,name=Expression"`
```

Comment:
need something more along the lines of https://github.com/cici37/kubernetes/blob/78d9388ecbbd7f13b200be021ce6531bd0a8c1f4/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go#L179-L231 here

I expect this documentation text to require the most effort of the whole API. It's the only place we have to explain to users what CEL variables will be accessible (xref https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3488-cel-admission-control/README.md#policy-definitions)

Reply:
I think a different variable name is appropriate here. `self` applies to whatever schema the validation rules are scoped to, but here we're offering a variable for the root of the object.